File: anb/leetcode/dynamic programming/1143. Longest Common Subsequence.py
class Solution(object):
    def rec(self, text1, text2, p1, p2):
        if p1 == len(text1) or p2 == len(text2):
            return 0
        if (p1, p2) in self.dp:
            return self.dp[p1, p2]
        self.dp[p1, p2] = 0
        
        if text1[p1] == text2[p2]:
            self.dp[p1, p2] = self.rec(text1, text2, p1 + 1, p2 + 1) + 1
    
        self.dp[p1, p2] = max(self.dp[p1, p2], self.rec(text1, text2, p1+1, p2), self.rec(text1, text2, p1, p2+1))
        return self.dp[p1, p2]
    
    def longestCommonSubsequence(self, text1, text2):
        """
        :type text1: str
        :type text2: str
        :rtype: int
        """
        self.dp = {}
        return self.rec(text1, text2, 0, 0)

File: gunkim/leetcode/dynamic programming/1143.Longest Common Subsequence.java
public class Main {
    public static void main(String[] args) {
        Solution solution = new Solution();
        System.out.println(solution.longestCommonSubsequence("abcde", "ace"));
    }
}

class Solution {
    public int longestCommonSubsequence(String text1, String text2) {
        int n = text1.length(), m = text2.length();
        int[][] dp = new int[n + 1][m + 1];

        for (int i = (n - 1); i >= 0; i--) {
            for (int j = (m - 1); j >= 0; j--) {
                if (text1.charAt(i) == text2.charAt(j)) {
                    dp[i][j] = 1 + dp[i + 1][j + 1];
                } else {
                    dp[i][j] = Math.max(dp[i][j + 1], dp[i + 1][j]);
                }
            }
        }
        return dp[0][0];
    }
}